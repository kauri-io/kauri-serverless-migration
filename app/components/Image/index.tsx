import styled, { css } from 'styled-components'
import { InView } from 'react-intersection-observer'

interface ImgProps {
    image: string
    className?: string
    height?: number | string
    width: number | string
    mobileHeight?: number | string
    mobileWidth?: number | string
    borderRadius?: string
    inView?: boolean
    borderTopLeftRadius?: string
    borderTopRightRadius?: string
    borderBottomLeftRadius?: string
    borderBottomRightRadius?: string
    children?: any
    overlay?: {
        color?: string
        opacity: number
    }
    asBackground?: boolean
    delay?: number
}

export const getURL = (
    url: string,
    height?: number | string,
    width?: number | string
) => {
    const heightParam = typeof height === 'number' ? height : null
    const widthParam = typeof width === 'number' ? width : null
    const getCDNURL = () => {
        if (heightParam && !widthParam) {
            return `https://${process.env.cloudImageId}.cloudimg.io/crop/2560x${heightParam}/webp/${url}`
        } else if (widthParam && !heightParam) {
            return `https://${process.env.cloudImageId}.cloudimg.io/widthParam/${widthParam}/webp/${url}`
        } else if (widthParam && heightParam) {
            return `https://${process.env.cloudImageId}.cloudimg.io/crop/${widthParam}x${heightParam}/webp/${url}`
        } else {
            return `https://${process.env.cloudImageId}.cloudimg.io/width/2560/webp/${url}`
        }
    }

    if (process.env.monolithApi && process.env.monolithApi.includes('uat')) {
        return getCDNURL()
    } else {
        return url
    }
}

const mobileDimensionsCSS = css<Pick<ImgProps, 'mobileHeight' | 'mobileWidth'>>`
    @media (max-width: ${props => props.theme.breakpoints[0]}) {
        height: ${props => props.mobileHeight};
        width: ${props => props.mobileWidth};
    }
`

const Img = styled.div<ImgProps>`
    height: ${props =>
        typeof props.height === 'number' ? `${props.height}px` : props.height};
    width: ${props =>
        typeof props.width === 'number' ? `${props.width}px` : props.width};
    ${props => props.borderRadius && `border-radius: ${props.borderRadius}`};
    background: ${props =>
        `${
            props.inView
                ? `url(${getURL(
                      props.image,
                      props.height,
                      props.width
                  )}) center center`
                : ``
        }`};
    background-size: cover;
    opacity: ${props => (props.inView ? 1 : 0)};
    transition: opacity ${props => props.delay || 0.4}s;
    ${props =>
        props.borderTopLeftRadius &&
        `border-top-left-radius: ${props.borderTopLeftRadius}`};
    ${props =>
        props.borderTopRightRadius &&
        `border-top-right-radius: ${props.borderTopRightRadius}`};
    position: ${props => (props.asBackground ? 'absolute' : 'relative')};

    ${props =>
        props.overlay &&
        `&:after {
      content: "";
      opacity: ${props.overlay.opacity};
      background: ${props.overlay.color || props.theme.colors.bgPrimary};
      height: 100%;
      width: 100%;
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      ${props.borderRadius && `border-radius: ${props.borderRadius}`};
      ${props.borderTopLeftRadius &&
          `border-top-left-radius: ${props.borderTopLeftRadius}`};
      ${props.borderTopRightRadius &&
          `border-top-right-radius: ${props.borderTopRightRadius}`};
      ${props.borderBottomLeftRadius &&
          `border-bottom-left-radius: ${props.borderBottomLeftRadius}`};
      ${props.borderBottomRightRadius &&
          `border-bottom-right-radius: ${props.borderBottomRightRadius}`};
    }`}

    ${props =>
        (props.mobileHeight || props.mobileWidth) && mobileDimensionsCSS};
`

const Image = (props: ImgProps) => (
    <InView rootMargin={'200px 0px'} triggerOnce={true}>
        {({ inView, ref }) => (
            <Img {...props} ref={ref} inView={inView}>
                {props.children}
            </Img>
        )}
    </InView>
)
export default Image
