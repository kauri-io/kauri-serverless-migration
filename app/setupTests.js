import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import theme from './lib/theme-config'
import { ThemeProvider } from 'styled-components'

Enzyme.configure({ adapter: new Adapter() })

const observeMock = {
    observe: () => null,
    disconnect: () => null, // maybe not needed
}

window.IntersectionObserver = () => observeMock

export const mountWithTheme = children =>
    mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const renderWithTheme = children =>
    render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const shallowWithTheme = children =>
    shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
