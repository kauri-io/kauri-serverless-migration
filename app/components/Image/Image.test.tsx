import Image, { getURL } from './index'
import { mountWithTheme } from '../../setupTests'

describe('components/Image', () => {
    const url = 'testURL'
    it('should match snapshot', () => {
        let props = {
            image: 'http://test.com/image.png',
            width: 100,
        }
        const wrapper = mountWithTheme(<Image {...props} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should return the plain url on development', () => {
        const result = getURL(url, 'development', 'cloudImageId', 10, 10)
        expect(result).toBe(url)
    })

    it('should return the default cloud image url if on production and no dimensions passed', () => {
        const result = getURL(url, 'production', 'cloudImageId')
        expect(result).toBe(
            `https://cloudImageId.cloudimg.io/width/2560/webp/${url}`
        )
    })

    it('should return the plain url on production with height and width', () => {
        const result = getURL(url, 'production', 'cloudImageId', 10, 10)
        expect(result).toBe(
            `https://cloudImageId.cloudimg.io/crop/10x10/webp/${url}`
        )
    })

    it('should return the plain url on production with height and no width', () => {
        const result = getURL(url, 'production', 'cloudImageId', 10)
        expect(result).toBe(
            `https://cloudImageId.cloudimg.io/crop/2560x10/webp/${url}`
        )
    })

    it('should return the plain url on production with width and no height', () => {
        const result = getURL(url, 'production', 'cloudImageId', undefined, 10)
        expect(result).toBe(`https://cloudImageId.cloudimg.io/10/webp/${url}`)
    })
})
