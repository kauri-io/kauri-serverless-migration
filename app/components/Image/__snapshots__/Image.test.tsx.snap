// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Image should match snapshot 1`] = `
<ThemeProvider
  theme={
    Object {
      "DEFAULT_CARD_WIDTH": 288,
      "MAIN_BODY_WIDTH": 1242,
      "bg": Object {
        "bgPrimary": "#1E2428",
        "bgSecondary": "#1E3D3B",
        "borderTextArea": "#979797",
        "contentBorder": "1px solid rgba(30, 36, 40, 0.19)",
        "disabledBackgroundColor": "#D6D6D6",
        "disabledTextColor": "#9B9B9B",
        "divider": "#EBEBEB",
        "errorRedColor": "#C03030",
        "hoverTextColor": "#108E72",
        "padding": "calc((100vw - \${props => props.theme.breakpoints[2]}) / 2)",
        "paddingTop": "2.5em",
        "primary": "#0BA986",
        "primaryColor": "#0BA986",
        "primaryDark": "#11856B",
        "primaryTextColor": "#1E2428",
        "secondary": "#11856B",
        "secondaryBlue": "rgb(81, 123, 162, 1)",
        "secondaryBlueDark": "rgb(81, 118, 142, 1)",
        "secondaryColor": "#1E3D3B",
        "secondaryTextColor": "#283035",
        "tertiaryBackgroundColor": "#F7F7F7",
        "textPrimary": "#1E2428",
        "white": "#FFFFFF",
      },
      "bgPrimary": "#1E2428",
      "bgSecondary": "#1E3D3B",
      "borderTextArea": "#979797",
      "breakpoints": Array [
        "500px",
        "52em",
        "1242px",
      ],
      "colors": Object {
        "bgPrimary": "#1E2428",
        "bgSecondary": "#1E3D3B",
        "borderTextArea": "#979797",
        "contentBorder": "1px solid rgba(30, 36, 40, 0.19)",
        "disabledBackgroundColor": "#D6D6D6",
        "disabledTextColor": "#9B9B9B",
        "divider": "#EBEBEB",
        "errorRedColor": "#C03030",
        "hoverTextColor": "#108E72",
        "padding": "calc((100vw - \${props => props.theme.breakpoints[2]}) / 2)",
        "paddingTop": "2.5em",
        "primary": "#0BA986",
        "primaryColor": "#0BA986",
        "primaryDark": "#11856B",
        "primaryTextColor": "#1E2428",
        "secondary": "#11856B",
        "secondaryBlue": "rgb(81, 123, 162, 1)",
        "secondaryBlueDark": "rgb(81, 118, 142, 1)",
        "secondaryColor": "#1E3D3B",
        "secondaryTextColor": "#283035",
        "tertiaryBackgroundColor": "#F7F7F7",
        "textPrimary": "#1E2428",
        "white": "#FFFFFF",
      },
      "contentBorder": "1px solid rgba(30, 36, 40, 0.19)",
      "disabledBackgroundColor": "#D6D6D6",
      "disabledTextColor": "#9B9B9B",
      "divider": "#EBEBEB",
      "errorRedColor": "#C03030",
      "fontSizes": Array [
        11,
        13,
        14,
        16,
        18,
        20,
        22,
        28,
        32,
      ],
      "fontWeight": Array [
        300,
        "normal",
        500,
        700,
      ],
      "hoverTextColor": "#108E72",
      "padContent": "
    max-width: 1242px;
    margin-left: auto;
    margin-right: auto;
    @media (min-width: 1242px) {
      min-width: 1242px;
    }
    @media (max-width: 1242px) {
      min-width: 90%;
    }
  ",
      "padding": "calc((100vw - 1242px) / 2)",
      "paddingTop": "2.5em",
      "primary": "#0BA986",
      "primaryColor": "#0BA986",
      "primaryDark": "#11856B",
      "primaryTextColor": "#1E2428",
      "secondary": "#11856B",
      "secondaryBlue": "rgb(81, 123, 162, 1)",
      "secondaryBlueDark": "rgb(81, 118, 142, 1)",
      "secondaryColor": "#1E3D3B",
      "secondaryTextColor": "#283035",
      "space": Array [
        0,
        10,
        15,
        30,
        60,
      ],
      "tertiaryBackgroundColor": "#F7F7F7",
      "textPrimary": "#1E2428",
      "white": "#FFFFFF",
    }
  }
>
  <Image
    image="http://test.com/image.png"
    width={100}
  >
    <InView
      rootMargin="200px 0px"
      threshold={0}
      triggerOnce={true}
    >
      <Image__Img
        image="http://test.com/image.png"
        inView={false}
        width={100}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Image__Img-sc-1jpuda7-0",
                "isStatic": false,
                "lastClassName": "iqWPmg",
                "rules": Array [
                  "height:",
                  [Function],
                  ";width:",
                  [Function],
                  ";",
                  [Function],
                  ";background:",
                  [Function],
                  ";background-size:cover;opacity:",
                  [Function],
                  ";transition:opacity ",
                  [Function],
                  "s;",
                  [Function],
                  ";",
                  [Function],
                  ";position:",
                  [Function],
                  ";",
                  [Function],
                  " ",
                  [Function],
                  ";",
                ],
              },
              "displayName": "Image__Img",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Image__Img-sc-1jpuda7-0",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={[Function]}
          image="http://test.com/image.png"
          inView={false}
          width={100}
        >
          <div
            className="Image__Img-sc-1jpuda7-0 iqWPmg"
            width={100}
          />
        </StyledComponent>
      </Image__Img>
    </InView>
  </Image>
</ThemeProvider>
`;
